Cython/Compiler/Dataclass.py

```diff
@@ -0,0 +1,594 @@
+    else:
+        code_lines.append("    return False")
+
+    code_lines = u"\n".join(code_lines)
+
+    return code_lines, {}, []
+
+def generate_eq_code(eq, node, fields):
+    if not eq:
+        return code_lines, {}, []
+    return generate_cmp_code("==", "__eq__", node, fields)
+
+def generate_order_code(order, node, fields):
+    if not order:
+        return "", {}, []
+    code_lines = []
+    placeholders = {}
+    stats = []
+    for op, name in [("<", "__lt__"),
+                     ("<=", "__le__"),
+                     (">", "__gt__"),
+                     (">=", "__ge__")]:
+        res = generate_cmp_code(op, name, node, fields)
+        code_lines.append(res[0])
+        placeholders.update(res[1])
+        stats.extend(res[2])
+    return "\n".join(code_lines), placeholders, stats
+
+def generate_hash_code(unsafe_hash, eq, frozen, node, fields):
+    hash_entry = node.scope.lookup_here("__hash__")
+    if hash_entry:
+        # TODO ideally assignment of __hash__ to None shouldn't trigger this
+        # but difficult to get the right information here
+        if unsafe_hash:
+            error(node.pos, "Request for dataclass unsafe_hash when a '__hash__' function"
+                  " already exists")
+        return "", {}, []
+    if not unsafe_hash:
+        if eq and not frozen:
+            return "", {}, [Nodes.SingleAssignmentNode(node.pos,
+                                        lhs = ExprNodes.NameNode(node.pos, name=EncodedString("__hash__")),
+                                        rhs = ExprNodes.NoneNode(node.pos))]
+        if not eq:
+            return
+
+    names = [ name for name, field in fields.items()
+                if (not field.is_initvar and
+                    (field.compare.value if field.hash.value is None else field.hash.value)) ]
+    if not names:
+        return "", {}, []  # nothing to hash
+
+    # make a tuple of the hashes
+    tpl = u", ".join(u"hash(self.%s)" % name for name in names )
+
+    # if we're here we want to generate a hash
+    code_lines = u"""def __hash__(self):
+    return hash((%s))
+""" % tpl
+
+    return code_lines, {}, []
+
+
+class GetTypeNode(ExprNodes.ExprNode):
+    # Tries to return a pytype_type if possible. However contains
+    # some fallback provision if it turns out not to resolve to a Python object
+    # Initialize with "entry"
+
+    subexprs = []
+
+    def __init__(self, entry):
+        super(GetTypeNode, self).__init__(entry.pos, entry=entry)
+
+    def analyse_types(self, env):
+        type = self.entry.type
+
+        if type.is_extension_type or type.is_builtin_type:
+            return ExprNodes.RawCNameExprNode(
+                self.pos, Builtin.type_type, type.typeptr_cname).analyse_types(env)
+        else:
+            names = None
+            py_name = type.py_type_name()
+            # int types can return "(int, long)"
+            if py_name:
+                names = [ name.strip("() ") for name in py_name.split(",") ]
```

Comment:
I've left this for now.

It sounds like it might be better as a `py_type_names` function that returns a list of strings (and maybe a convenience `py_types_for_instance_check` function that turns it to a tuple if needed)